src/cooldown.rs

```diff
@@ -34,14 +38,32 @@ pub struct CooldownConfig {
     pub __non_exhaustive: (),
 }
 
+/// Trait to allow loading custom [`CooldownConfig`]s based on the current [`CooldownContext`]
+#[async_trait::async_trait]
+pub trait CooldownConfigProvider<U, E>: Debug + Send + Sync {
```

Comment:
Did you try `for<'a> Fn(&'a CooldownContext) -> BoxFuture<'a, Result<CooldownConfig, E>>`?

Reply:
I did see some similar uses elsewhere in the code. I was hitting some issues with that, but that's likely just because I don't *really* understand what `for<'a>` does (it means the trait bound needs to work for all lifetimes `'a` but I haven't internalized what the implications of that are).

I took another crack at it and totally got it (mostly) working. Although functions passed in through `#[poise::command(cooldown_provider)]` can't be `async` and need to manually handle the `BoxFuture` stuff. Still messing around to see if I can get that working nicely.